### DISCUSSION TITLE
Prevent or cancel navigation?

### DISCUSSION THREAD

**Discussion by gleclair** [lepoco/wpfui#1290] [Q&A]
First off, many thanks to @pomianowski and other contributors for creating and maintaining this solution,  your work is very much appreciated!!
My question is:   Is there a way for a view model to disable or cancel navigation?   If a user is in the middle of editing, I want them to either save or discard changes before moving to another screen.  Maybe I could disable the navigation menu when editing starts, or wire into an event to cancel it?  Hopefully I'm missing something easy, any suggestions are appreciated.

**Comment by gleclair** [comment]
I was able to use CommunityToolKit.MVVM.Messaging so each viewmodel can send a message to the MainWindowViewModel when it's busy and wants to block navigation.  MainWindowViewModel will then set its CanNavigate property to false.   My MainWindow code behind wires to the RootNavigation_Navigating event:
`  private void RootNavigation_Navigating(NavigationView sender, NavigatingCancelEventArgs args)
  {
      if(!ViewModel.CanNavigate) args.Cancel = true;
  }`
This solution works alright, but there's a few issues:

1. There's no visual indication that navigation is not currently allowed. 
2. The navigation seems to only be partially cancelled, for example the view model for the requested page is actually constructed.   

Has anyone got a suggestion to improve this?   Thanks!

**Reply by gleclair** [reply]
So nobody else has this need?   Usually that means I'm missing something obvious, so I thought I'd ask again.  
I know Fluent designs like this emphasize minimal user friction and use auto-save as much as possible, but my app has a lot of data entry - we're creating new database records and we need to get a complete set of data before saving.      Modal dialogs for data gathering work but they're cumbersome.  
I've started customizing the NavigationService and NavigationView in the source, if I get that all working cleanly I'll create a PR.   But in the meantime I'm still looking for a better suggestion.

**Reply by bhdryzgn** [reply]
I too have searched a solution for this not yet found a cleanest way. If you come up with any you may assume "clean" solution please share. Thank you in advance.

**Reply by RichardD2** [reply]
I'm also looking for a solution - preferable one which would allow an `async` handler on the current page to decide whether or not to cancel the navigation, since every option for asking the user (short of falling back to the system `MessageBox`) requires `async`.

But even without an `async` option, as far as I can see, there's no way from the `Navigating` event to know which page you're navigating *from*, since the `Page` property on the `NavigatingCancelEventArgs` is the *destination* page instance. And as you mentioned, that means the destination page instance has already been created when the event fires.

**Comment by Florin-Purice** [comment]
@gleclair This doesn't address your original problem, just a modification of your event handler solution using styles instead.

This disables the navigation buttons, so the navigation should be fully cancelled.
There still is no visual tell when the navigation is disabled (should be doable by setting the background or something but idc too much about this).

I still rely on IMessenger for the page viewmodel to communicate with MainWindow viewmodel . Put this inside ```NavigationView``` in MainWindow.xaml
```xaml
<ui:NavigationView.Resources>
    <Style TargetType="{x:Type ui:NavigationViewItem}" BasedOn="{StaticResource NavigationViewItemDefaultStyle}">
        <Setter Property="IsEnabled" Value="{Binding ViewModel.CanNavigate, Mode=OneWay}" />
    </Style>
</ui:NavigationView.Resources>
```
